### PR TITLE
PT-1329: Resolves dynamic property metadata for incoming requests if it is missing.

### PR DESCRIPTION
**Pull request by tatarincev**
Added `DynamicObjectPropertyJsonConverter`, which resolves metadata for all incoming dynamic properties if it is missing.
Now you can update dynamic properties values through  API without specifying meta-data (property id, value type etc)
POST http://localhost:10645/api/members
```Json
.....
    "dynamicProperties": [
        {
            "values": [
                {
                   
                    "value": "s",
                    "valueId": "ea8a1fa1-4a48-49a0-9378-3f45a1747225",
                    "propertyName": "a"
                },
                {
                    "value": "a",
                    "valueId": "45563845-647a-4bd0-8f5b-e720c774fa92"
                }
            ],
    
            "name": "a"         
        }
    ],
....
```

### CODE REVIEW

File: src/VirtoCommerce.Platform.Core/JsonConverters/DynamicObjectPropertyJsonConverter.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+        public override bool CanConvert(Type objectType)
+        {
+           return typeof(IHasDynamicProperties).IsAssignableFrom(objectType);
+        }
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            var result = base.ReadJson(reader, objectType, existingValue, serializer);
+            if(result != null && result is IHasDynamicProperties hasDynProp)
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Core/JsonConverters/DynamicObjectPropertyJsonConverter.cs:26]: Redundant check.
```suggestion
            if (result is IHasDynamicProperties hasDynamicProperties)
```

File: src/VirtoCommerce.Platform.Core/JsonConverters/DynamicObjectPropertyJsonConverter.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        private readonly IDynamicPropertyMetaDataResolver _metadataResolver;
+        public DynamicObjectPropertyJsonConverter(IDynamicPropertyMetaDataResolver metadataResolver)
+        {
+            _metadataResolver = metadataResolver;
+        }
+
+        public override bool CanConvert(Type objectType)
+        {
+           return typeof(IHasDynamicProperties).IsAssignableFrom(objectType);
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Core/JsonConverters/DynamicObjectPropertyJsonConverter.cs:20]: Fix styling, please.

File: src/VirtoCommerce.Platform.Data/DynamicProperties/DynamicPropertyMetaDataResolver.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+               for (var skip = _pageSize; skip < totalCount; skip += _pageSize)
+               {
+                   criteria.Skip = skip;
+                   searchResult = await _searchService.SearchDynamicPropertiesAsync(criteria);
+                   result.AddRange(searchResult.Results);
+               }
+               return result.Distinct().ToDictionary(x => $"{x.ObjectType}__{x.Name}", StringComparer.InvariantCultureIgnoreCase).WithDefaultValue(null);
+
+           });
+           return dict[$"{objectType}__{propertyName}"];
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Data/DynamicProperties/DynamicPropertyMetaDataResolver.cs:61]: After `.WithDefaultValue(null)` we can get `NullReferenceException`.

File: src/VirtoCommerce.Platform.Data/DynamicProperties/DynamicPropertyMetaDataResolver.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                throw new ArgumentNullException(nameof(objectType));
+            }
+            if (propertyName == null)
+            {
+                throw new ArgumentNullException(nameof(propertyName));
+            }
+
+            var cacheKey = CacheKey.With(GetType(), nameof(GetByNameAsync));
+            var dict = await _memoryCache.GetOrCreateExclusiveAsync(cacheKey, async (cacheEntry) =>
+           {
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Data/DynamicProperties/DynamicPropertyMetaDataResolver.cs:38]: Fix styling, please.

**Review by Vectorfield4 — CHANGES_REQUESTED** [review]
Fix code smells, please.

**Review by akak1977 — APPROVED** [review]
Seems ok, need to throw out smells.

File: src/VirtoCommerce.Platform.Data/DynamicProperties/DynamicPropertyMetaDataResolver.cs (hunk 3 above)
Comment (tatarincev) [reply src/VirtoCommerce.Platform.Data/DynamicProperties/DynamicPropertyMetaDataResolver.cs]: Please explain in which case we can get a null reference?

File: src/VirtoCommerce.Platform.Core/DynamicProperties/DynamicPropertiesExtensions.cs [hunk 5]
```diff
@@ -1,13 +1,58 @@
...
+            if (metaDataResolver == null)
+            {
+                throw new ArgumentNullException(nameof(metaDataResolver));
+            }
+
+            var hasDynProps = owner.GetFlatObjectsListWithInterface<IHasDynamicProperties>();
+            foreach (var hasDynProp in hasDynProps)
+            {
+                await ResolveMetaDataAsync(hasDynProp, metaDataResolver);
+            }
```
[the remark this answers is not in the document]
Comment (tatarincev) [reply src/VirtoCommerce.Platform.Core/DynamicProperties/DynamicPropertiesExtensions.cs:28]: Do these tasks in parallel can be potential leads for resource-consuming and threads starvation under heavy load. I prefer to left the sequential execution.

**Review by eMazeika — APPROVED** [review]

**Review by Vectorfield4 — APPROVED** [review]

**Pull request merged**